### DISCUSSION TITLE
Add support for TN3270

### DISCUSSION THREAD

**Discussion by blackbit42** [jquast/telnetlib3#143] [Ideas]
Hi @jquast 

Have you ever looked at TN3270?
Could you imagine to add support for it?
Relevant RFCs are:

TN3270 Current Practices
https://datatracker.ietf.org/doc/html/rfc1576 Jan 1994

TN3270 Extensions for LUname and Printer Selection
https://datatracker.ietf.org/doc/html/rfc1646 Jul 1994

TN3270 Enhancements
https://datatracker.ietf.org/doc/html/rfc2355 Jun 1998

TN3270E Service Location and Session Balancing
https://datatracker.ietf.org/doc/html/rfc3049 Jan 2001

There are a few projects floating around that implement TN3270 servers in various languages, including python, though they all are rather nonchalant regarding telnet negotiation and focus on the application part.
It would be great to do TN3270 by the book.

**Comment by jquast** [comment]
Hello @blackbit42, yes I was interested in TN3270, I did read those RFC's years ago when I implemented the others, I think I tried to implement some of it but later removed them, because I couldn't test it. I don't have any IBM servers clients to test with, this stuff is all commercial and costly.

If somebody suggests a PR that they've tested I'd certainly accept it.

**Comment by blackbit42** [comment]
Testing is easy with http://x3270.bgp.nu/ actually.

**Comment by jquast** [comment]
and which open source IBM server would I test it with?

**Comment by blackbit42** [comment]
Oh, for client connections you mean.
I am not aware of any open source implementations by IBM or any other big brand vendor.
There are mainframe OS images on archive.org that have been used successfully to run TN3270 servers. I am not going to advice using them though.
There are also quite a number of publicly accessible servers on the internet.
```x rutadmin.rutgers.edu 992 TLS
x tammvs1.tamu.edu 992 TLS
x tammvst.tamu.edu 992 TLS TN3270E cert name mismatch
x teague-tammvs1.tamu.edu TLS TN3270E cert name mismatch
x 174.47.13.112 23 plain
x 65.196.87.141 23 plain TN3270E
x 47.19.64.73 23 plain TN3270E
x 47.19.64.74 23 plain TN3270E
x 80-122.206-83.static-ip.oleane.fr plain TN3270E
x vm.se.lsoft.com plain TN3270E
x 81.80.140.164 plain TN3270E <error message>
x 173-165-108-185-Illinois.hfc.comcastbusiness.net plain TN3270E
x ecommsvr.csclogic.com plain TN3270E
x 199.190.128.4 plain TN3270E
x 172-126-122-51.lightspeed.rlghnc.sbcglobal.net plain TN3270E
x 47.19.64.74 plain TN3270E
x IVY.UMIS.UPENN.EDU 992 TLS TN3270E
x IVYTEST.UMIS.UPENN.EDU 992 TLS TN3270E
x iron.umis.upenn.edu 992 TLS TN3270E
x 192.86.32.17 623 plain TN3270E
? ace.amadeus.net 992 TLS
x 77.92.104.14 623 plain
x 192.86.32.153 623 plain
x ctrek.dal-ebis.ihost.com 623 plain
x mntn.state.mn.us 2443 TLS TN3270E
```

But I guess you are looking for something else.

**Comment by jquast** [comment]
Just to follow up that I would want TN3270 support to be bidirectional -- both client and server, so that it can be used with itself to provide code coverage. I can't easily write only client or server support against 3rd party code, it would have to be complete enough to operate as a unit test like other protocols supported in the library

**Comment by blackbit42** [comment]
Makes sense.
There are open source client and server implementations readily available for testing.
Prime example of TN3270 client is https://x3270.bgp.nu/
Prime example of TN3270 server is https://github.com/SDL-Hercules-390/hyperion

**Comment by jquast** [comment]
nice profile photo by the way, you might like to try my demo at ``telnet 1984.ws``

**Comment by jquast** [comment]
A little bit of support was added for next release #118 to allow --encoding=cp037 / ebcdic to also allow negotiation of ``NEW_ENVIRON`` variables in the same encoding (ebcdic), this partially allows for correct negotiation but I didn't go any further.
